subsys/lorawan/services/lorawan_services.c

```diff
@@ -0,0 +1,181 @@
+	uint8_t port;
+	/* sufficient space for up to 3 answers (max 6 bytes each) */
+	uint8_t data[18];
+	uint8_t len;
+	/* absolute ticks when this message should be scheduled */
+	int64_t ticks;
+	sys_snode_t node;
+};
+
+K_THREAD_STACK_DEFINE(thread_stack_area, CONFIG_LORAWAN_SERVICES_THREAD_STACK_SIZE);
+
+/*
+ * The services need a dedicated work queue, as the LoRaWAN stack uses the system
+ * work queue and gets blocked if other LoRaWAN messages are sent and processed from
+ * the system work queue in parallel.
+ */
+static struct k_work_q services_workq;
+
+static struct k_work_delayable uplink_work;
+
+/* single-linked list (with pointers) and array for implementation of priority queue */
+static struct service_uplink_msg messages[10];
+static sys_slist_t msg_list;
+static struct k_sem msg_sem;
+
+static void uplink_handler(struct k_work *work)
+{
+	struct service_uplink_msg msg_copy;
+	struct service_uplink_msg *first;
+	sys_snode_t *node;
+	int err;
+
+	ARG_UNUSED(work);
+
+	/* take semaphore and create a copy of the next message */
+	k_sem_take(&msg_sem, K_FOREVER);
+
+	node = sys_slist_get(&msg_list);
+	if (node == NULL) {
+		goto out;
+	}
+
+	first = CONTAINER_OF(node, struct service_uplink_msg, node);
+	msg_copy = *first;
+	first->used = false;
+	sys_slist_remove(&msg_list, NULL, &first->node);
+
+	/* semaphore must be given back before calling lorawan_send */
+	k_sem_give(&msg_sem);
+
+	err = lorawan_send(msg_copy.port, msg_copy.data, msg_copy.len, LORAWAN_MSG_UNCONFIRMED);
+	if (!err) {
+		LOG_DBG("Message sent to port %d", msg_copy.port);
+	} else {
+		LOG_ERR("Sending message to port %d failed: %d",
+			msg_copy.port, err);
+	}
+
+	/* take the semaphore again to schedule next uplink */
+	k_sem_take(&msg_sem, K_FOREVER);
+
+	node = sys_slist_peek_head(&msg_list);
+	if (node == NULL) {
+		goto out;
+	}
+	first = CONTAINER_OF(node, struct service_uplink_msg, node);
+	k_work_reschedule_for_queue(&services_workq, &uplink_work,
+		K_TIMEOUT_ABS_TICKS(first->ticks));
+
+out:
+	k_sem_give(&msg_sem);
+}
+
+static inline void insert_uplink(struct service_uplink_msg *msg_new)
+{
+	struct service_uplink_msg *msg_prev;
+
+	if (sys_slist_is_empty(&msg_list)) {
+		sys_slist_append(&msg_list, &msg_new->node);
+	} else {
+		int count = 0;
+
+		SYS_SLIST_FOR_EACH_CONTAINER(&msg_list, msg_prev, node) {
+			count++;
+			if (msg_prev->ticks <= msg_new->ticks) {
+				break;
+			}
+		}
+		if (msg_prev != NULL) {
+			sys_slist_insert(&msg_list, &msg_prev->node, &msg_new->node);
+		} else {
+			sys_slist_append(&msg_list, &msg_new->node);
+		}
+	}
+}
+
+int lorawan_services_schedule_uplink(uint8_t port, uint8_t *data, uint8_t len, k_timeout_t timeout)
+{
+	struct service_uplink_msg *next;
+	int64_t timeout_abs_ticks;
+
+	if (len > sizeof(messages[0].data)) {
+		LOG_ERR("Uplink payload for port %u too long: %u bytes", port, len);
+		LOG_HEXDUMP_ERR(data, len, "Payload: ");
+		return -EFBIG;
+	}
+
+	timeout_abs_ticks = k_uptime_ticks() + timeout.ticks;
```

Comment:
This doesn't work if the provided timeout is already absolute.

Reply:
Fixed. I've replaced it with an `uint32_t timeout` in milliseconds instead of the opaque type to make it safer.